designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,505 @@
+
+```js
+const plugin = {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+    // can't include the plugin here because it must reference `plugin`
+	configs: {},
+};
+
+Object.assign(plugin.configs, {
+	recommended: {
+
+        // now we can reference plugin
+		plugins: { json: plugin },
+		rules: {
+			"json/no-duplicate-keys": "error",
+			"json/no-empty-keys": "error",
+		},
+	},
+});
+
+export default plugin;
+```
+
+Here, we are hardcoding the namespace `json` even though that might not be the namespace that the user assigns to this plugin. This is something we can now address with the use of `extends` because we have the ability to alter the config before inserting it.
+
+Instead of using a hardcoded plugin namespace, plugins can instead use `#` to indicate that they'd like to have the plugin itself included and referenced using the namespace the user assigned. For example, we could rewrite the JSON plugin like this:
+
+```js
+export default {
+	meta: {
+		name: "@eslint/json",
+		version: "0.6.0", // x-release-please-version
+	},
+	languages: {
+		json: new JSONLanguage({ mode: "json" }),
+		jsonc: new JSONLanguage({ mode: "jsonc" }),
+		json5: new JSONLanguage({ mode: "json5" }),
+	},
+	rules: {
+		"no-duplicate-keys": noDuplicateKeys,
+		"no-empty-keys": noEmptyKeys,
+	},
+
+	configs: {
+        recommended: {
+            plugins: { "#": null },
+            rules: {
+                "#/no-duplicate-keys": "error",
+                "#/no-empty-keys": "error",
+            },
+        },
+    },
+};
+```
+
+In an `eslint.config.js` file:
+
+```js
+import json from "@eslint/json";
+
+export default [
+    {
+        plugins: {
+            j: json
+        },
+        files: ["**/*.json"],
+        extends: ["j/recommended"],
+        rules: {
+            "j/no-empty-keys": "warn"
+        }
+    }
+];
+```
+
+Here, the user has assigned the namespace `j` to the JSON plugin. When ESLint loads this config, it sees `extends` with a value of `"j/recommended"`. It looks up the plugin referenced by `j` and then the config named `"recommended"`. It then sees that there's a plugin entry for `"#"` and replaces that with an entry for `j` that points to the JSON plugin. The next step is to look through the config for all the `#` references and replace that with `j` so the references are correct. 
+
+### Implementation Details
+
+The implementation of this feature requires the following changes:
+
+1. Enable nested arrays in `FlatConfigArray`
+1. Create a new `ConfigExtender` class that encapsulates the functionality for extending configs.
+1. Update `FlatConfigArray` to use `ConfigExtender` inside of the `normalize()` and `normalizeSync()` methods.
```

Comment:
Since `FlatConfigArray` is not something that's exported (see discussion in https://github.com/eslint/eslint/issues/18619) this would not be something that the [`config-inspector`](https://github.com/eslint/config-inspector) can make use of but something that it would need to re-implement, which would be a lot of extra work and a risk of divergence / differences in implementation.

I would strongly suggest that the non-exporting of `FlatConfigArray` is reconsidered if all this is added there.

An option could be that the `ConfigExtender` is added as part of `@eslint/config-array` instead and that `ConfigArray` rather than `FlatConfigArray` is updated to use it in `normalize()`, and that it behaves like nested arrays: Its something one opts into / out of for a config array

Thoughts @antfu?

Reply:
Yeah, the fact I love the flat config format also makes config-inspector possible is that it's very transparent, where you can import the config file as a plain js module without eslint to get all the information. What I understand is that we now want to move to a bit of the middle ground to have ESLint interop the config a bit to provide better DX, which I would be love to see.

In that regard, in order to keep the inspector working (where we still present the final resolved flat configs to users), I agree with @voxpelli that we do need ESLint to expose those API to handle the config resolutions to the final config array.

Reply:
That's good feedback re: `FlatConfigArray` and a good argument for adding this functionality into `@eslint/config-array` directly.

I'm still hesitant to export `FlatConfigArray` itself because we are still making changes and I don't want to be tied to that API by exposing it.